src/harness/test_harness_objects.py

```diff
@@ -0,0 +1,319 @@
+          'highFrequency': heartbeat_request['operationParam']['operationFrequencyRange']['highFrequency']
+        }
+      }
+    }
+    return grant_request
+
+  def performHeartbeatAndUpdateGrants(self):
+    """Does heartbeat request for all active grants and updates the grants.
+
+    The following steps are performed by this function
+      1. Send heartbeat request on all grant of CBSD at once
+      2. If the heartbeat response contains operationParam then those grants
+         are relinquished.
+      3. New grant requests with the suggested operationParam values in the
+         heartbeat response is sent. Failures are ignored.
+      4. Heartbeat is performed again for all grants. Failures are ignored.
+    """
+    # Heartbeat on all active grants
+    heartbeat_requests = []
+    heartbeat_requests, heartbeat_responses = self.heartbeatRequestForAllActiveGrants()
+    relinquish_requests = []
+    grant_requests = []
+
+    # Construct relinquishment and grant for the heart beat response which have operation param
+    for heartbeat_response, heartbeat_request in zip(heartbeat_responses, heartbeat_requests):
+        if heartbeat_response['response']['responseCode'] == Grant.ResponseCodes.TERMINATED_GRANT.value:
+           if 'operationParam' in heartbeat_response:
+             # Construct grant request for Terminated grants
+             grant_requests.append(self._constructGrantRequests(heartbeat_request))
+           else:
+             logging.error('Invalid Response since operation param not present in case of Terminated Grant')
+
+        elif heartbeat_response['response']['responseCode'] != 0:
+          # If any heartbeat contain suggested operation params,construct relinquish requests
+          if 'operationParam' in heartbeat_response:
+            relinquish_request = {
+                'cbsdId': heartbeat_response['cbsdId'],
+                'grantId': heartbeat_response['grantId']
+            }
+            relinquish_requests.append(relinquish_request)
+
+            # Construct grant request for relinquished grants
+            grant_requests.append(self._constructGrantRequests(heartbeat_request))
+
+            # Delete the grant object
+            del self.cbsd_objects[heartbeat_response['cbsdId']].grant_objects[heartbeat_response['grantId']]
+
+    # Perform relinquishment since operation param present in heartbeat response
+    if len(relinquish_requests):
+      relinquish_requests_wrap = {
+        'relinquishmentRequest': relinquish_requests
+      }
+      relinquish_responses = self.testcase._sas.Relinquishment(relinquish_requests_wrap,
+                                        self.ssl_cert, self.ssl_key)['relinquishmentResponse']
+      for relinquish_response in relinquish_responses:
+        if relinquish_response['response']['responseCode'] != 0:
+          logging.error('Relinquishment request for the grantId= %s failed',relinquish_response['grantId'])
```

Comment:
Not just log an error, raise an exception. The primary reason a relinquishment could fail is if the test is sending invalid relinquishment requests (in which case the test result is invalid and the bug should be fixed), otherwise something has gone horribly wrong and the SAS UUT is inconsistent with the state expected by the test (in which case any subsequent requests have unpredictable responses, and it could be an error in either the SAS UUT or the test code).

Reply:
Agreed. We have added assert statements to verify that the   relinquishment is successful.